pkg/bootconfig/zipconfig.go

```diff
@@ -0,0 +1,130 @@
+package bootconfig
+
+import (
+	"archive/zip"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+
+	"github.com/systemboot/systemboot/pkg/crypto"
+	"golang.org/x/crypto/ed25519"
+)
+
+// memoryZipReader is used to unpack a zip file from a byte sequence in memory.
+type memoryZipReader struct {
+	Content []byte
+}
+
+func (r *memoryZipReader) ReadAt(p []byte, offset int64) (n int, err error) {
+	cLen := int64(len(r.Content))
+	if offset > cLen {
+		return 0, io.EOF
+	}
+	if cLen-offset >= int64(len(p)) {
+		n = len(p)
+		err = nil
+	} else {
+		err = io.EOF
+		n = int(int64(cLen) - offset)
+	}
+	copy(p, r.Content[offset:int(offset)+n])
+	return n, err
+}
+
+// FromZip tries to extract a boot configuration from a ZIP file after verifying
+// its signature with the provided public key file. The signature is expected to
+// be appended to the ZIP file and have fixed length `ed25519.SignatureSize` .
+// The returned string argument is the temporary directory where the files were
+// extracted, if successful.
+// No decoder (e.g. JSON, ZIP) or other function parsing the input file is called
+// before verifying the signature.
+func FromZip(filename string, pubkeyfile *string) (*Manifest, string, error) {
+	// load the whole zip file in memory - we need it anyway for the signature
+	// matching.
+	// TODO refuse to read if too big?
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, "", err
+	}
+	zipbytes := data
+	// Load the public key
+	if pubkeyfile != nil {
+		zipbytes = data[:len(data)-ed25519.SignatureSize]
+		pubkey, err := crypto.LoadPublicKeyFromFile(*pubkeyfile)
+		if err != nil {
+			return nil, "", err
+		}
+
+		// Load the signature.
+		// The signature is appended to the zip file and has length
+		// `ed25519.SignatureSize`. We read these bytes from the end of the file and
+		// treat them as the attached signature.
+		signature := data[len(data)-ed25519.SignatureSize : len(data)]
+		if len(signature) != ed25519.SignatureSize {
+			return nil, "", fmt.Errorf("Short read when reading signature: want %d bytes, got %d", ed25519.SignatureSize, len(signature))
+		}
+
+		// Verify the signature against the public key and the zip file bytes
+		if ok := ed25519.Verify(pubkey, zipbytes, signature); !ok {
+			return nil, "", fmt.Errorf("Invalid ed25519 signature for file %s", filename)
+		}
+	} else {
+		log.Printf("No public key specified, the ZIP file will be unpacked without verification")
```

Comment:
You can append any arbitrary content to a ZIP file and it would still be extracted correctly, it's a known aspect of ZIP (and other file formats) that enables "features" like this. I can add a comment to explain it